### PR TITLE
Move Pickles.compile to a separate Pickles.Compile module

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR builds upon #12630, factoring out the implementation of `Pickles.compile` into its own module, so that the adversarial override helper in `Pickles.compile_with_wrap_main_override_promise` can be used for testing outside of `pickles.ml`.

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Comment by mrmr1993** [comment]
!ci-build-me

File: src/lib/pickles/pickles.ml [hunk 1]
```diff
@@ -3935,7 +2893,7 @@ module Make_str (_ : Wire_types.Concrete) = struct
 
         module Proof = (val p)
 
-        let example1 =
+        let _example1 =
```
Comment (jspada) [inline src/lib/pickles/pickles.ml:2896]: are these getting moved?

**Review by jspada — APPROVED** [review]
Looks good to me!  Just moving things around.

File: src/lib/pickles/pickles.ml (hunk 1 above)
Comment (mrmr1993) [reply src/lib/pickles/pickles.ml:2896]: I turned on warnings to help me deduplicate things; these values are unused, so these renamings squash the warnings.

File: src/lib/pickles/compile.mli [hunk 2]
```diff
@@ -0,0 +1,329 @@
+module Scalar_challenge = Scalar_challenge
+module Endo = Endo
```
Comment (rbonichon) [inline src/lib/pickles/compile.mli:2]: Do you think we could remove these aliases to identically named local modules?

File: src/lib/pickles/compile.mli [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+open Core_kernel
+open Async_kernel
+open Pickles_types
+open Hlist
+module Common = Common
```
Comment (rbonichon) [inline src/lib/pickles/compile.mli:7]: Same here: I have not found any matches for Common within the signature

**Review by rbonichon — APPROVED** [review]
Overall LGTM. Only minor issues

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Pull request merged**